### ISSUE TITLE
False positive for using nullable method with strict Contract("null -> null; !null -> !null") as a function !null -> !null

### ISSUE THREAD

**Issue by unv-unv** [uber/NullAway#1502]
We have a method (toUpperCase) which returns null only if its argument is null. Otherwise it always returns not null. So it ensures non-null result for non-null argument and can be safely passed as lambda to be used in non-null context. But NullAway doesn't detect this situation and shows error:

```java
@NullMarked
public class Test {
    @Contract("null -> null; !null -> !null")
    public @Nullable String toUpperCase(@Nullable String s) {
        return s == null ? null : s.toUpperCase(Locale.US);
    }

    public String[] map(String[] strings, Function<String, String> mapper) {
        String[] result = new String[strings.length];
        for (int i = 0, n = strings.length; i < n; i++) {
            result[i] = mapper.apply(strings[i]);
        }
        return result;
    }

    public String[] capitalizeWords(String[] words) {
        // BUG: [NullAway] referenced method returns @Nullable, but functional interface method java.util.function.Function.apply(T) returns @NonNull
        return map(words, StringUtil::toUpperCase);
    }
}
```

NullAway 0.13.1, Error Prone 2.48.0

**Comment by msridhar** [comment]
Very sorry for the late response on this.  This is a valid issue; just want to spell out my reasoning so I don't forget.  `toUpperCase` is a "poly-null" function; in the Checker Framework Nullness Checker, we would write `public @PolyNull String toUpperCase(@PolyNull String s) { ... }` without the `@Contract` annotation.  This means that whenever it is passed a non-null String, it is guaranteed to return a non-null String.  So, intuitively a reference is safe to pass to a `Function<String,String>`, since via that type, only non-null Strings can be passed in.

All that said, I'm not 100% sure how much this rule generalizes beyond the `Function` type, particularly in the presence of wildcards, etc.  So we'll have to be careful in how broad a fix we put in.
